zerver/lib/streams.py

```diff
@@ -1005,3 +1008,33 @@ def get_subscribed_private_streams_for_user(user_profile: UserProfile) -> QueryS
         .filter(subscribed=True)
     )
     return subscribed_private_streams
+
+
+@transaction.atomic
+def check_update_all_streams_active_status(days: int = 180) -> int:
+    date_days_ago = timezone_now() - timedelta(days=days)
+
+    active_stream_ids = (
+        Message.objects.filter(date_sent__gte=date_days_ago, recipient__type=Recipient.STREAM)
+        .values_list("recipient__type_id", flat=True)
+        .distinct()
+    )
+    streams_to_mark_inactive = Stream.objects.filter(is_recently_active=True).exclude(
+        id__in=active_stream_ids
+    )
+
+    # Send events to notify the users about the change in the stream's active status per realm.
+    for stream in streams_to_mark_inactive:
+        event = dict(
+            type="stream",
+            op="update",
+            property="is_recently_active",
+            value=False,
+            stream_id=stream.id,
+            name=stream.name,
+        )
+        send_event_on_commit(stream.realm, event, active_user_ids(stream.realm_id))
+
+    count = streams_to_mark_inactive.update(is_recently_active=False)
+
+    return count
```

Comment:
As mentioned above, this needs to be split into a loop over each realm, with a transaction inside that loop.

If we're going to leave this every-3-months (which I do not suggest), then we should make this send single event with a list of streams, so we don't potentially cause a huge bolus of events.

Reply:
Done.